### ISSUE TITLE
[🐛 Bug]: Tests failing since docker grid update to version 4.19

### ISSUE THREAD

**Issue by d4v1dk0n7** [SeleniumHQ/docker-selenium#2175]
### What happened?

Hi all,

I've updated the docker selenium grid to latest version 4.19, and now our tests are not working.
Due to this I had to revert the grid to the previous version 4.18.1

Any advise would be appreciated..
Thank you!

### Command used to start Selenium Grid with Docker (or Kubernetes)

```shell
version: "3.8"

services:
    vsts_agent:
      build:
        context: .
        dockerfile: docker/Dockerfile.vsts
      image: vsts_agent
      environment:
        - AZP_URL=https://xxxx
        - AZP_TOKEN=xxxx
        - AZP_POOL=xxxx
      volumes:
        - /srv:/srv
      scale: 4
      restart: always

    selenium-hub:
      build:
        context: .
        dockerfile: docker/Dockerfile.hub
      container_name: selenium-hub
      ports:
        - 4442:4442
        - 4443:4443
        - 4444:4444
      environment:
        - SE_SESSION_REQUEST_TIMEOUT=1200
      ulimits:
        nofile: 32768
      restart: always

    node-docker:
      build:
        context: .
        dockerfile: docker/Dockerfile.node-docker
      container_name: node-docker
      shm_size: 2gb
      volumes:
        - ./assets:/opt/selenium/assets
        - ./NodeDocker/config.toml:/opt/bin/config.toml
        - /var/run/docker.sock:/var/run/docker.sock
      depends_on:
        - selenium-hub
      environment:
        - SE_EVENT_BUS_HOST=selenium-hub
        - SE_EVENT_BUS_PUBLISH_PORT=4442
        - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
        - SE_NODE_MAX_SESSIONS=16
        - SE_NODE_SESSION_TIMEOUT=1200
        - SE_OPTS=--session-timeout 1200
        - SE_START_XVFB=false
        - SE_START_VNC=false
      ulimits:
        nofile: 32768
      restart: always

    chrome:
      build:
          context: .
          dockerfile: docker/Dockerfile.chrome
      container_name: chrome
      shm_size: 2gb
      depends_on:
        - selenium-hub
      ulimits:
        nofile: 32768
      restart: always

    edge:
      build:
          context: .
          dockerfile: docker/Dockerfile.edge
      container_name: edge
      shm_size: 2gb
      depends_on:
        - selenium-hub
      ulimits:
        nofile: 32768
      restart: always

    firefox:
      build:
          context: .
          dockerfile: docker/Dockerfile.firefox
      container_name: firefox
      shm_size: 2gb
      depends_on:
        - selenium-hub
      ulimits:
        nofile: 32768
      restart: always
```

### Relevant log output

```shell
Error Message:
   OneTimeSetUp: System.InvalidOperationException : Could not start a new session. Could not start a new session. non-positive contentLength: 0 
Host info: host: '6dbe86d5cb61', ip: '172.19.0.7'
Build info: version: '4.19.0', revision: '5f9cec8963'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-97-generic', java.version: '17.0.10'
Driver info: driver.version: unknown
Build info: version: '4.19.0', revision: '5f9cec8963'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '5.15.0-97-generic', java.version: '17.0.10'
Driver info: driver.version: unknown (SessionNotCreated)
  Stack Trace:
     at OpenQA.Selenium.WebDriver.UnpackAndThrowOnError(Response errorResponse, String commandToExecute)
   at OpenQA.Selenium.WebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.WebDriver.StartSession(ICapabilities capabilities)
```

### Operating System

Ubuntu 22.04

### Docker Selenium version (image tag)

4.19

**Comment by VietND96** [comment]
Can you also share the list of capabilities that are set from binding? It seems the error relates to it
```
   at OpenQA.Selenium.WebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.WebDriver.StartSession(ICapabilities capabilities)
```

**Comment by d4v1dk0n7** [comment]
Hi!

These are the capabilities we are using:

                    var options = new ChromeOptions();
                    options.AddArgument("--window-size=1920,1080");
                    options.AddArguments("--headless=new");
                    options.AddArguments("--no-sandbox");
                    options.AddArguments("--disable-gpu");
                    options.PageLoadStrategy = PageLoadStrategy.Eager;
                    options.SetLoggingPreference(LogType.Browser, LogLevel.All);                   
                    webDriver = new RemoteWebDriver(new Uri("http://xxxxx:4444/"), options.ToCapabilities(), TimeOut);

**Comment by diemol** [comment]
I am not able to reproduce this with the images we provide. You are using custom Docker images. What happens when you use the images we provide?

If it still fails, can you please provide the full test including dependencies?

**Comment by VietND96** [comment]
I think something relates to this change https://github.com/SeleniumHQ/selenium/commit/7c1fae379f502e78772b81f389cfecb441392081
for the method `options.ToCapabilities()` in dotnet binding

It can be an instance of `SetLoggingPreference` not in the data type `Dictionary<string, object>`

**Comment by diemol** [comment]
@nvborisenko do you think that is the issue?

**Comment by d4v1dk0n7** [comment]
> I am not able to reproduce this with the images we provide. You are using custom Docker images. What happens when you use the images we provide?
> 
> If it still fails, can you please provide the full test including dependencies?

@diemol I'm using the official selenium docker images

**Comment by diemol** [comment]
Why do you have a build file?
```
    chrome:
      build:
          context: .
          dockerfile: docker/Dockerfile.chrome
```

**Comment by d4v1dk0n7** [comment]
> Why do you have a build file?
> 
> ```
>     chrome:
>       build:
>           context: .
>           dockerfile: docker/Dockerfile.chrome
> ```

Because we have to copy some resources to the images, for the tests.
This is the content of the build file:

FROM selenium/standalone-chrome:4.18.1

COPY 1-Infrastructure/QA.Resources/Files /tmp/Resources

ENV TZ=Europe/Madrid
RUN sudo ln -snf /usr/share/zoneinfo/$TZ /etc/localtime

**Comment by nvborisenko** [comment]
@d4v1dk0n7 what if logging are enabled? https://www.selenium.dev/documentation/webdriver/troubleshooting/logging

> Could not start a new session. non-positive contentLength: 0

This is something returned by the server.

**Comment by d4v1dk0n7** [comment]
> @d4v1dk0n7 what if logging are enabled? https://www.selenium.dev/documentation/webdriver/troubleshooting/logging
> 
> > Could not start a new session. non-positive contentLength: 0
> 
> This is something returned by the server.

Hi @nvborisenko,
What do you mean? Sorry I don't understand..

**Comment by nvborisenko** [comment]
In .net code insert the following before creating new RemoteWebDriver instance:
```csharp
Internal.Logging.Log.SetLevel(Internal.Logging.LogEventLevel.Trace);
```

Then please share console output.

**Comment by d4v1dk0n7** [comment]
> In .net code insert the following before creating new RemoteWebDriver instance:
> 
> ```cs
> Internal.Logging.Log.SetLevel(Internal.Logging.LogEventLevel.Trace);
> ```
> 
> Then please share console output.

Hi @nvborisenko, unfortunately I won't be able to try this until next tuesday (out of the office)...
Please keep the issue opened until then!

Thanks.

**Comment by nvborisenko** [comment]
I tried your code with my local docker based on https://github.com/SeleniumHQ/docker-selenium/blob/trunk/docker-compose-v3.yml compose file - no issues, works fine.

**Comment by VietND96** [comment]
It looks like the issue is with a specific node-docker (dynamic grid). In another comment also reported https://github.com/SeleniumHQ/docker-selenium/issues/2154#issuecomment-2025210340
In CI is missing tests to ensure dynamic grid works against regression changes.
I will add a test, identify the root cause, make the test pass, and release a new image tag.

**Comment by VietND96** [comment]
I suspected it could be JRE 17 in the base image, but even though I reverted, the issue still occurs. So, I think it could be an issue in upstream Selenium. @diemol, do you know what the issue in the logs below could be?

```bash
19:22:13.932 DEBUG [JdkHttpClient.execute0] - Executing request: (POST) /v1.41/images/create
19:22:13.931 DEBUG [HttpClientImpl$SelectorManager.run] - [HttpClient-1-SelectorManager] [1s 205ms] HttpClientImpl(1) next timeout: 0
19:22:13.932 DEBUG [HttpClientImpl$SelectorManager.run] - [HttpClient-1-SelectorManager] [1s 205ms] HttpClientImpl(1) next expired: 1199068
19:22:13.932 DEBUG [HttpClientImpl$SelectorManager.run] - [HttpClient-1-SelectorManager] [1s 205ms] HttpClientImpl(1) Next deadline is 3000
19:22:13.932 DEBUG [JdkHttpClient.execute0] - Ending request (POST) /v1.41/images/create in 0ms
19:22:13.932 DEBUG [SeleniumSpanExporter$1.lambda$export$4] - SpanData{spanContext=ImmutableSpanContext{traceId=568cd549092fb543c58cf8a78f876458, spanId=5c39ea56a5cc69ec, traceFlags=01, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=true}, parentSpanContext=ImmutableSpanContext{traceId=00000000000000000000000000000000, spanId=0000000000000000, traceFlags=00, traceState=ArrayBasedTraceState{entries=[]}, remote=false, valid=false}, resource=Resource{schemaUrl=null, attributes={service.name="unknown_service:java", telemetry.sdk.language="java", telemetry.sdk.name="opentelemetry", telemetry.sdk.version="1.36.0"}}, instrumentationScopeInfo=InstrumentationScopeInfo{name=default, version=null, schemaUrl=null, attributes={}}, name=httpclient.execute, kind=INTERNAL, startEpochNanos=1711740133932100937, endEpochNanos=1711740133932641559, attributes=AttributesMap{data={span.kind=client, http.target=/v1.41/images/create, http.method=POST}, capacity=128, totalAddedValues=3}, totalAttributeCount=3, events=[], totalRecordedEvents=0, links=[], totalRecordedLinks=0, status=ImmutableStatusData{statusCode=UNSET, description=}, hasEnded=true}
19:22:13.933 ERROR [Bootstrap.runMain] - Error during execution
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.openqa.selenium.grid.Bootstrap.runMain(Bootstrap.java:77)
	at org.openqa.selenium.grid.Bootstrap.main(Bootstrap.java:70)
Caused by: org.openqa.selenium.grid.config.ConfigException: java.lang.reflect.InvocationTargetException
	at org.openqa.selenium.grid.config.MemoizedConfig.getClass(MemoizedConfig.java:119)
	at org.openqa.selenium.grid.node.config.NodeOptions.getNode(NodeOptions.java:181)
	at org.openqa.selenium.grid.node.httpd.NodeServer.createHandlers(NodeServer.java:126)
	at org.openqa.selenium.grid.node.httpd.NodeServer.asServer(NodeServer.java:185)
	at org.openqa.selenium.grid.node.httpd.NodeServer.execute(NodeServer.java:247)
	at org.openqa.selenium.grid.TemplateGridCommand.lambda$configure$4(TemplateGridCommand.java:122)
	at org.openqa.selenium.grid.Main.launch(Main.java:83)
	at org.openqa.selenium.grid.Main.go(Main.java:56)
	at org.openqa.selenium.grid.Main.main(Main.java:41)
	... 6 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.openqa.selenium.grid.config.ClassCreation.callCreateMethod(ClassCreation.java:51)
	at org.openqa.selenium.grid.config.MemoizedConfig.lambda$getClass$4(MemoizedConfig.java:104)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(Unknown Source)
	at org.openqa.selenium.grid.config.MemoizedConfig.getClass(MemoizedConfig.java:99)
	... 14 more
Caused by: java.lang.IllegalArgumentException: non-positive contentLength: 0
	at java.net.http/java.net.http.HttpRequest$BodyPublishers.fromPublisher(Unknown Source)
	at org.openqa.selenium.remote.http.jdk.JdkHttpMessages.notChunkingBodyPublisher(JdkHttpMessages.java:124)
	at org.openqa.selenium.remote.http.jdk.JdkHttpMessages.createRequest(JdkHttpMessages.java:82)
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute0(JdkHttpClient.java:371)
	at org.openqa.selenium.remote.http.AddSeleniumUserAgent.lambda$apply$0(AddSeleniumUserAgent.java:42)
	at org.openqa.selenium.remote.http.Filter.lambda$andFinally$1(Filter.java:55)
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.execute(JdkHttpClient.java:352)
	at org.openqa.selenium.remote.tracing.TracedHttpClient.execute(TracedHttpClient.java:54)
	at org.openqa.selenium.docker.v1_41.PullImage.apply(PullImage.java:62)
	at org.openqa.selenium.docker.v1_41.V141Docker.getImage(V141Docker.java:79)
	at org.openqa.selenium.docker.Docker.lambda$getImage$0(Docker.java:50)
	at java.base/java.util.Optional.map(Unknown Source)
	at org.openqa.selenium.docker.Docker.getImage(Docker.java:50)
	at org.openqa.selenium.grid.node.docker.DockerOptions.lambda$loadImages$5(DockerOptions.java:259)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Exception in thread "Thread-0" java.lang.NullPointerException: Cannot invoke "org.openqa.selenium.grid.node.Node.getStatus()" because "this.node" is null
	at org.openqa.selenium.grid.node.httpd.NodeServer.lambda$new$0(NodeServer.java:78)
	at java.base/java.lang.Thread.run(Unknown Source)
2024-03-29 19:22:14,277 INFO exited: selenium-grid-docker (exit status 1; not expected)
```

**Comment by d4v1dk0n7** [comment]
Hi all,

Any news regarding this issue?
Can I help in any way by my side?

Thanks!

**Comment by diemol** [comment]
Still needs to be fixed.

**Comment by d4v1dk0n7** [comment]
Hi @diemol,

Do you know something regarding this issue?
It's not fixed yet?

Thanks!

**Comment by diemol** [comment]
We haven't had time to fix it yet. The linked commit helps to understand what changes need to be made. Would you like to help?

**Comment by d4v1dk0n7** [comment]
> We haven't had time to fix it yet. The linked commit helps to understand what changes need to be made. Would you like to help?

I'd like to help but I'm afraid I don't have the necessary skills...
Thanks for the info!!

**Comment by VietND96** [comment]
The issue is fixed on upstream repo, and will be released in `4.20.0`
In the meantime, you can try with image tag `nightly` for preview - `selenium/node-docker:nightly`
